mypyc/irbuild/function.py

```diff
@@ -907,15 +922,65 @@ def gen_dispatch_func_ir(
     return dispatch_func_ir
 
 
-def sort_with_subclasses_first(
-    impls: List[RegisterImplInfo]
-) -> Iterator[RegisterImplInfo]:
+def load_singledispatch_registry(builder: IRBuilder, fitem: FuncDef, line: int) -> Value:
+    name = get_registry_identifier(fitem)
+    return builder.add(LoadStatic(dict_rprimitive, name, builder.module_name, line=line))
+
+
+def initialize_all_singledispatch_dictionaries(builder: IRBuilder) -> None:
+    singledispatch_funcs = builder.singledispatch_impls.keys()
+    for func in singledispatch_funcs:
+        main_func_name = singledispatch_main_func_name(func.name)
+        initialize_singledispatch_dispatch_dictionary(builder, func, main_func_name)
+
+
+def initialize_singledispatch_dispatch_dictionary(
+    builder: IRBuilder,
+    fitem: FuncDef,
+    main_func_name: str
+) -> None:
+    line = fitem.line
+    func_obj = load_func(builder, main_func_name, fitem.fullname, line)
+
+    impls = builder.singledispatch_impls[fitem]
+    native_impls = []
+
+    loaded_object_type = builder.load_module_attr_by_fullname('builtins.object', line)
+    registry_dict = builder.builder.make_dict([(loaded_object_type, func_obj)], line)
+
+    for dispatch_type, impl in impls:
+        if is_decorated(builder, impl):
```

Comment:
Hm, would it be possible to add decorated functions to the registry immediately after they are decorated, instead of in a centralized way like this?

Reply:
Actually, we could probably also put the native implementations into the registry at the point in running the toplevel that would correspond to the function as well, and that would correspond better to the real cpython semantics of when things get registered.